packages/web3-eth-abi/src/coders/base/array.ts

```diff
@@ -0,0 +1,120 @@
+/*
+This file is part of web3.js.
+
+web3.js is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+web3.js is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
+*/
+import { AbiError } from 'web3-errors';
+import { AbiParameter } from 'web3-types';
+import { uint8ArrayConcat } from 'web3-utils';
+// eslint-disable-next-line import/no-cycle
+import { decodeParamFromAbiParameter, encodeNumber, encodeParamFromAbiParameter } from '.';
+import { DecoderResult, EncoderResult } from '../types.js';
+import { extractArrayType, isDynamic, WORD_SIZE } from '../utils.js';
+import { decodeNumber } from './number.js';
+import { encodeDynamicParams } from './utils.js';
+
+export function encodeArray(param: AbiParameter, values: unknown): EncoderResult {
+	if (!Array.isArray(values)) {
+		throw new AbiError('Expected value to be array', { abi: param, values });
+	}
+	const { size, param: arrayItemParam } = extractArrayType(param);
+	const encodedParams = values.map(v => encodeParamFromAbiParameter(arrayItemParam, v));
+	const dynamic = size === -1;
+	const dynamicItems = encodedParams.length > 0 && encodedParams[0].dynamic;
+	if (!dynamic && values.length !== size) {
+		throw new AbiError("Given arguments count doesn't match array length", {
```

Comment:
What do you mean? its the same error instance as before. Adding new error messages is not a breaking chnage

Reply:
I think this is not just a new error message. The code used to throw an error for this case, and for others, but with a different message and a different structure, right?
The old message was according to the old deleted code:
```throw new AbiError(`Parameter encoding error`, err as Error);``` And the innerError was the error object we get from the encoder.

Reply:
Code still throws AbiError error and our api doesn't guarantee inner error type will stay the same (It will be of type Error | Error[] if present). It's user's fault if he expected something that typescript types did not guarantee.

Reply:
Thanks @mpetrunic,
Yes, it is not guaranteed by typescript types here. But still the user can even use JavaScript which has no types. I am thinking here about the actual strings and values that the user used to get and might already has `if`s checks on. And if we think it is the user's fault to depend on types we did not specify, changing them is still considered a breaking change, as far as I know. 

So, as far as I know, if we do not maintain to provide the exact message string and the exact internal properties for every error, as we used to, then we are introducing a breaking change. And yes we can add more properties to the error object. But we are not supposed to remove or update the string or a property value, without pumping the major version.  

And yes, our API has been used to guarantee the same message and properties's values for the innerError message because the early used library is following semver.

And I think keeping the error messages and properties, as they were, is possible even though it might be a bit tricky. 

Please correct me if changing an error message, or a value of an internal property inside an error object, is not considered a breaking change.

Thanks,

Reply:
Afaik, error messages are implementation details intended for the developer's eyes. No one should rely on their exact content or even worse make an if/else statement based on their content (fixing a typo in an error message would then be a breaking change). If the developer intended for you to handle those errors, errors would either be different instances or contain error codes along with the list of error codes the function might throw.